### PR TITLE
fix: favor extended array toStringTag where available

### PR DESCRIPTION
**Pull request by dvlsg**
Discussion in #49 

Still pending any specific tests to cover this scenario. If you'd like me to add tests as part of this pull request, let me know where you'd like them. (`new-ecmascript-types.js`?)

### CODE REVIEW

**Comment by keithamus** [comment]
Hey @dvlsg, thanks for this. Looks great!

Tests would definitely be ideal before we merge. We have toStringTag tests [here in test/index.js#L227-261](https://github.com/chaijs/type-detect/blob/f7dcaf02f23a7ad2628fd52efc5510d2fc3cc3b3/test/index.js#L227-L261). If you could add Arrays there, that'd be great. Thanks!

**Comment by dvlsg** [comment]
Can do.

**Comment by dvlsg** [comment]
Is it safe for me to assume that `test/index.js` can be run in environments that may not support classes?

Considering adding checks like found [here](https://github.com/chaijs/type-detect/blob/master/test/new-ecmascript-types.js#L15-L23). Are those worthwhile in this case, or just extra noise?

**Comment by keithamus** [comment]
@dvlsg yes indeed. The tests will be run in IE9, PhantomJS, etc. classes are not a given. Feel free to wrap it in the style we have done, such as in the example you linked.

**Comment by dvlsg** [comment]
Hm. Thoughts on the Node v4.3 case? It's coming out as `[object Array]` with the code I have. I can take a closer look later on, as well.

File: index.js [hunk 1]
```diff
@@ -95,7 +95,9 @@ module.exports = function typeDetect(obj) {
    *   array literal      x 22,479,650 ops/sec ±0.96% (81 runs sampled)
    */
   if (isArrayExists && Array.isArray(obj)) {
-    return 'array';
+    if (symbolToStringTagExists === false || typeof obj[Symbol.toStringTag] === 'undefined') {
```
Comment (keithamus) [inline index.js]: These ifs could be combined for terseness.

File: test/index.js [hunk 2]
```diff
@@ -258,6 +261,26 @@ describe('Generic', function () {
...
+      '}',
+    ].join('\n');
+    try {
+      eval(extendedArrayBody); // eslint-disable-line no-eval
+      extendedArraySupported = true;
+    } catch (error) {
+      extendedArraySupported = false;
+    }
+
+    itIf(extendedArraySupported)('extended array', function () {
```
Comment (keithamus) [inline test/index.js]: Node 4.3 doesn't support Symbol.toStringTag so you'll need to add something like:

```js
itIf(typeof Symbol !== 'undefined' && typeof Symbol.toStringTag !== 'undefined' && extendedArraySupported)(...
```

Comment (keithamus) [inline test/index.js]: oh wait actually this is inside the sham isn't it? Forget the above then

File: test/index.js [hunk 3]
```diff
@@ -258,6 +261,26 @@ describe('Generic', function () {
       assert(type(obj) === 'foo', 'type(obj) === "foo"');
     });
 
+    var extendedArraySupported = false;
+    var extendedArrayBody = [
+      'class ExtendedArray extends Array {',
+      '  get [Symbol.toStringTag]() {',
+      '    return "ExtendedArray";',
```
Comment (keithamus) [inline test/index.js]: I'm not sure this should be a getter, `[Symbol.toStringTag]` is meant to be a function

File: test/index.js [hunk 4]
```diff
@@ -249,7 +253,6 @@ describe('Generic', function () {
       Object.prototype.toString = originalObjectToString; // eslint-disable-line no-extend-native
```
Comment (keithamus) [inline test/index.js]: I have a feeling you'll need to change the sham behaviour to test for getter properties, right now it only asserts that the `[Symbol.toStringTag]` prop is a function (which in the case of your class, it isn't)

**Comment by dvlsg** [comment]
All makes sense to me, here's another swing at it, might not be 100% there yet.

Couple notes about what I did:

* I think you're right about `Symbol.toStringTag` being a function. I know it *works* as a getter (for whatever reason), but a function is probably the more right way, so I made that swap.
* I swapped the logic in `index.js` to check if `Symbol` exists instead of if `Symbol.toStringTag` exists. I'm hoping that's okay for all 4.0 branches, like the [announcement](https://nodejs.org/en/blog/release/v4.0.0/) suggests. Seems to be okay for 4.3, at any rate. Otherwise, the logic I originally had was failing the `symbolToStringTagExists` check and sending back `array` instead of dropping through to give back `extendedarray`. The change allows the unit tests to pass, but it's a bit awkward at best, especially considering that I believe this would be checking `typeof array[undefined]` on node 4.3.  I'm all ears if you have a better suggestion for covering that scenario.
* I combined the if statements per your suggestion, and took a stab at what sort of formatting you'd want for a multiline if (assuming you'd even *want* a multiline if...) since putting it all on one line made the linter sad. (I think it was at `123` characters, so close!)

**Comment by meeber** [comment]
I believe the `symbolToStringTagExists` check is necessary anywhere `.toStringTag` is used. There are some browser versions that support `Symbol` but not `Symbol.toStringTag` which means this:

```js
(symbolExists === false || typeof obj[Symbol.toStringTag] === 'undefined')
```

...ends up meaning this:

```js
(symbolExists === false || typeof obj['undefined'] === 'undefined')
```

...which can lead to some funky behavior, especially when the `obj` being tested is a Chai Assertion. See #25

**Comment by dvlsg** [comment]
I agree, it's definitely odd.

As far as I can tell, `global.Symbol.toStringTag` only makes it into the scope of `index.js` *after* `symbolToStringTagExists` is calculated (which makes sense).

```js
// test/index.js
// using node v4.3
// index.js utilizing `symbolToStringTagExists`
console.log(Symbol.toStringTag); //=> '__@@toStringTag__'
console.log(Object.prototype.toString.call(extendedArray); //=> '[object ExtendedArray]'
console.log(type(extendedArray)); //=> 'array'
```

We could *technically* swap the if statement to be something like this:

```js
// ...
if (
  isArrayExists &&
  Array.isArray(obj) &&
  (typeof Symbol === 'undefined' || typeof Symbol.toStringTag === 'undefined' || typeof obj[Symbol.toStringTag] === 'undefined')
) {
  return 'array';
}
```

But then we're running those `typeof`'s *every* time, which seems a bit silly to get this specific case to pass inside the toString sham section for only node 4.3. I'd love to say how much of a performance hit it would be, but my laptop just died on me in the middle of `npm run bench`. :)

I agree that I don't like the `typeof obj['undefined']` stuff, though, for sure. Maybe we're better off trying to make this specific unit test skip v4.3 like v0.10 since `Symbol.toStringTag` isn't technically available at that point? Idk.

File: index.js [hunk 5]
```diff
@@ -94,7 +94,11 @@ module.exports = function typeDetect(obj) {
    * Post:
    *   array literal      x 22,479,650 ops/sec ±0.96% (81 runs sampled)
    */
-  if (isArrayExists && Array.isArray(obj)) {
+  if (
+    isArrayExists &&
+    Array.isArray(obj) &&
+    (symbolExists === false || typeof obj[Symbol.toStringTag] === 'undefined')
+  ) {
```
Comment (keithamus) [inline index.js]: Typically the precedent is that if we have long conditions, they get assigned to a var which is more readable. For example here:

```js
var symbolToStringTagExists = (symbolExists === false || typeof obj[Symbol.toStringTag] === 'undefined');
if (isArrayExists && Array.isArray(obj) && symbolToStringTagExists)
```

Would be ideal. Both lines <120 and the if is more readable as the conditions are given names with semantic meaning.

**Comment by keithamus** [comment]
> As far as I can tell, global.Symbol.toStringTag only makes it into the scope of index.js after symbolToStringTagExists is calculated (which makes sense).

Yes... this makes things more complicated. You could sham Symbol.toStringTag before requiring the main lib from inside the tests, which would get around this problem - but it might make other tests fail. I guess give it a try 😄

**Comment by meeber** [comment]
Now that I think about it, I ran into this same problem when working on #25. 

An argument could be made that shamming stuff in the tests is making things more difficult than they need to be. Instead, there could be separate tests depending on if `Symbol` and/or `Symbol.toStringTag` are defined, and both situations would be covered via testing multiple node versions in Travis.

**Comment by dvlsg** [comment]
When I get back to my laptop (hopefully later tonight), I'll try lifting to toStringTag sham stuff toward the top of the file / before the main `require`, if that's acceptable. If it doesn't affect any of the other unit tests, that might work as a (potentially temporary) fix.

I also like the idea of splitting the tests into separate environments based on `Symbol` and `Symbol.toStringTag` availability, but I might stay away from that in this PR (if possible, assuming the `toStringTag` lifting works) to keep the issues separate.

**Comment by dvlsg** [comment]
Hah. This is turning into *way* more trouble than I thought it would be! Here's where it's at:

Lifting the `toStringTag` sham stuff seems to work as expected, *except* it only works when I run `mocha ./test/index.js` directly, because otherwise a different test file will load `./index.js` first, which will calculate `symbolToStringTag` exists before the sham has a chance to run and then cache that `require` to feed to the other tests. :disappointed:

Considering invalidating the [require cache](http://stackoverflow.com/questions/9210542/node-js-require-cache-possible-to-invalidate)? That would technically work, I think.

**Comment by meeber** [comment]
Highly recommend that we hold on this PR, create a new PR to refactor tests so that no shamming is involved, then rebase this PR with the changes, thus making writing tests significantly easier (as well as positioning us better for the future).

**Comment by keithamus** [comment]
Invalidating the require cache in a `before` step would be a relatively good way to solve this. It's a bit of a cludge but definitely solves the problem.

**Comment by dvlsg** [comment]
Well, cludge number one (the `require` reload) had to be wrapped in cludge number two (checking to see if `require.resolve()` is available) since phantomjs [doesn't support it](https://github.com/ariya/phantomjs/pull/11972). I suppose we could put together something that doesn't need to use `require.resolve()` to get the id / path of the module in `require.cache`, but since the tests seem to work in phantomjs without needing to invalidate the require cache and reload, I opted for this for now.

Also, this commit log is getting fairly noisy, since I keep pushing partial failures. :\ Sounds like we may be holding off on a merge until a test refactor? But it's probably worth squashing to reduce the noise. Github allows squashing before merging, [right](https://help.github.com/articles/about-pull-request-merge-squashing/)? Otherwise, I can do it manually.

**Comment by keithamus** [comment]
Yeah, let's pause progress on this PR for a while, and we'll get the tests reliable in a different PR. If you have some ideas on how to get the tests reliable, please feel free to raise a new PR, otherwise I'll investigate over next weekend.

**Comment by keithamus** [comment]
@dvlsg if you would like to revisit this issue, I think it'd be much simpler to solve with the latest `master`. Should just be a case of moving the `toStringTag` check above `Array.isArray`.

**Comment by keithamus** [comment]
Just FYI @dvlsg I'm closing this one, as I've rebased it and made it available at #95, and updated the tests to ensure they're working. Thanks for your hard work on this!

**Comment by dvlsg** [comment]
Sweet! I got very distracted by other work (apologies!), so I'm glad to see it's making it in! Thank you for picking it up.

**Pull request closed without merging**